tests/federation_room_upgrade_test.go

```diff
@@ -0,0 +1,123 @@
+package tests
+
+import (
+	"reflect"
+	"testing"
+	"time"
+
+	"github.com/tidwall/gjson"
+
+	"github.com/matrix-org/complement/internal/b"
+	"github.com/matrix-org/complement/internal/client"
+	"github.com/matrix-org/complement/internal/match"
+	"github.com/matrix-org/complement/internal/must"
+)
+
+// NOTE: this test falls under a "should" guarantee from the spec: https://spec.matrix.org/v1.4/client-server-api/#server-behaviour-16
+// > When a user joins the new room, the server SHOULD automatically transfer/replicate some of the user’s personalized settings such as notifications, tags, etc.
+//
+// sytest: $user_type user has push rules copied to upgraded room
+func TestRoomUpgradeCopiesPushRules(t *testing.T) {
+	deployment := Deploy(t, b.BlueprintFederationTwoLocalOneRemote)
+	defer deployment.Destroy(t)
+
+	alice := deployment.Client(t, "hs1", "@alice:hs1")
+	bob := deployment.Client(t, "hs1", "@bob:hs1")
+	charlie := deployment.Client(t, "hs2", "@charlie:hs2")
+
+	doTestWith := func(t *testing.T, user *client.CSAPI) {
+		roomID := alice.CreateRoom(t, map[string]interface{}{
+			"preset": "public_chat",
+		})
+
+		user.JoinRoom(t, roomID, []string{"hs1"})
+		alice.MustSyncUntil(t, client.SyncReq{}, client.SyncJoinedTo(user.UserID, roomID))
```

Comment:
Why is _alice_ syncing? Presumably this should be `user.MustSyncUntil`?

Reply:
If `user` asserts that it has joined the room, the join might not have propagated to `alice`'s homeserver, so when `alice` asserts it, it must've registered on both `user` and `alice`'s homeserver, which means its safe to assume this going forward, whereas its not when only `user` sees it.

It might look small, but it could result in a flake if `alice`'s homeserver is just slow enough.

Reply:
> when alice asserts it, it must've registered on both user and alice's homeserver,

while this is true, it's not obvious to the reader at first glance.

Can we either
- have a comment along the lines of your own above, or
- change (2) to have `user` wait for the join, and then  change alice's sync on +43 to wait for `user`'s join before upgrading the room.